File: benchmark-framework-source/src/main/java/com/benchmark/report/JsonExporter.java
package com.benchmark.report;

import com.benchmark.core.BenchmarkResult;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * Exports benchmark results to a structured JSON file with no external libraries.
 */
public final class JsonExporter {

    private final String outputDir;

    public JsonExporter(String outputDir) {
        this.outputDir = outputDir;
    }

    public String export(Map<String, List<BenchmarkResult>> resultsByTask) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        sb.append("  \"generatedAt\": \"").append(Instant.now()).append("\",\n");
        sb.append("  \"framework\": \"Java Concurrency Benchmark Framework\",\n");
        sb.append("  \"tasks\": [\n");

        int ti = 0;
        for (Map.Entry<String, List<BenchmarkResult>> entry : resultsByTask.entrySet()) {
            if (ti++ > 0) sb.append(",\n");
            sb.append("    {\n");
            sb.append("      \"taskName\": \"").append(escape(entry.getKey())).append("\",\n");
            sb.append("      \"concurrencyLevels\": [\n");

            int ri = 0;
            for (BenchmarkResult r : entry.getValue()) {
                if (ri++ > 0) sb.append(",\n");
                sb.append("        {\n");
                sb.append("          \"concurrency\": ").append(r.getConcurrencyLevel()).append(",\n");
                sb.append("          \"totalIterations\": ").append(r.getTotalIterations()).append(",\n");
                sb.append("          \"totalDurationMs\": ").append(r.getTotalDurationMs()).append(",\n");
                sb.append("          \"throughputOpsPerSec\": ").append(round(r.getThroughput())).append(",\n");
                sb.append("          \"avgLatencyMs\": ").append(round(r.getAvgLatencyMs())).append(",\n");
                sb.append("          \"medianLatencyMs\": ").append(round(r.getMedianLatencyMs())).append(",\n");
                sb.append("          \"p95LatencyMs\": ").append(round(r.getP95LatencyMs())).append(",\n");
                sb.append("          \"p99LatencyMs\": ").append(round(r.getP99LatencyMs())).append(",\n");
                sb.append("          \"avgCpuPercent\": ").append(round(r.getAvgCpuPercent())).append(",\n");
                sb.append("          \"peakCpuPercent\": ").append(round(r.getPeakCpuPercent())).append(",\n");
                sb.append("          \"avgMemoryMB\": ").append(round(r.getAvgMemoryMB())).append(",\n");
                sb.append("          \"peakMemoryMB\": ").append(round(r.getPeakMemoryMB())).append("\n");
                sb.append("        }");
            }
            sb.append("\n      ]\n");
            sb.append("    }");
        }
        sb.append("\n  ]\n}\n");

        File out = new File(outputDir, "benchmark-results.json");
        try (Writer w = new FileWriter(out, StandardCharsets.UTF_8)) {
            w.write(sb.toString());
        }
        return out.getAbsolutePath();
    }

    private static double round(double v) { return Math.round(v * 100.0) / 100.0; }
    private static String escape(String s) { return s.replace("\\","\\\\").replace("\"","\\\""); }
}

File: python/benchmark/report/json_exporter.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Mapping

from benchmark.core.benchmark_result import BenchmarkResult


class JsonExporter:
    def __init__(self, output_dir: str) -> None:
        self._output_dir = Path(output_dir)

    def export(self, results_by_task: Mapping[str, List[BenchmarkResult]]) -> str:
        self._output_dir.mkdir(parents=True, exist_ok=True)
        out_path = self._output_dir / "benchmark-results-python.json"

        tasks = []
        for task_name, results in results_by_task.items():
            levels = []
            for r in results:
                levels.append(
                    {
                        "concurrency": r.concurrency_level,
                        "totalIterations": r.total_iterations,
                        "totalDurationMs": r.total_duration_ms,
                        "throughputOpsPerSec": _round(r.get_throughput()),
                        "avgLatencyMs": _round(r.get_avg_latency_ms()),
                        "medianLatencyMs": _round(r.get_median_latency_ms()),
                        "p95LatencyMs": _round(r.get_p95_latency_ms()),
                        "p99LatencyMs": _round(r.get_p99_latency_ms()),
                        "avgCpuPercent": _round(r.avg_cpu_percent),
                        "peakCpuPercent": _round(r.peak_cpu_percent),
                        "avgMemoryMB": _round(r.avg_memory_mb),
                        "peakMemoryMB": _round(r.peak_memory_mb),
                    }
                )
            tasks.append({"taskName": task_name, "concurrencyLevels": levels})

        payload = {
            "generatedAt": datetime.now(timezone.utc).isoformat(),
            "framework": "Python Concurrency Benchmark Framework",
            "language": "python",
            "tasks": tasks,
        }

        out_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        return str(out_path.resolve())


def _round(v: float) -> float:
    return round(v * 100.0) / 100.0
